cloudinit/sources/DataSourceGCE.py

```diff
@@ -88,13 +89,50 @@ def __init__(self, sys_cfg, distro, paths):
 
     def _get_data(self):
         url_params = self.get_url_params()
-        network_context = noop()
         if self.perform_dhcp_setup:
-            network_context = EphemeralDHCPv4(
-                self.distro,
-                self.fallback_interface,
-            )
-        with network_context:
+            candidate_nics = net.find_candidate_nics()
+            if DEFAULT_PRIMARY_INTERFACE in candidate_nics:
+                candidate_nics.remove(DEFAULT_PRIMARY_INTERFACE)
+                candidate_nics.insert(0, DEFAULT_PRIMARY_INTERFACE)
+            LOG.debug("Looking for the primary NIC in: %s", candidate_nics)
+            assert (
+                len(candidate_nics) >= 1
+            ), "The instance has to have at least one candidate NIC"
+            for candidate_nic in candidate_nics:
+                network_context = EphemeralDHCPv4(
+                    self.distro,
+                    iface=candidate_nic,
+                )
+                try:
+                    with network_context:
+                        try:
+                            ret = util.log_time(
+                                LOG.debug,
+                                "Crawl of GCE metadata service",
+                                read_md,
+                                kwargs={
+                                    "address": self.metadata_address,
+                                    "url_params": url_params,
+                                },
+                            )
+                        except Exception:
+                            LOG.debug(
+                                "Error fetching IMD with candidate NIC %s",
+                                candidate_nic,
+                                exc_info=True,
+                            )
+                            continue
```

Comment:
Thanks for the supplemental handling.

Let's drop the exc_info=True here. We don't need/want to inject tracebacks in cloud-init logs for this as it'll be hard to read for minimal benefit.

Can we instead just append the exception str(e) here? 

```suggestion
                        except Exception as e:
                            LOG.debug(
                                "Error fetching IMD with candidate NIC %s: %s",
                                candidate_nic,
                                e,
                            )
                            continue
```

Also since we know the likelihood of seeing HTTPConnectionPool type errors for non-primary devices like `2023-06-05 11:30:52,610 - DataSourceGCE.py[DEBUG]: url instance/id raised exception HTTPConnectionPool(host='metadata.google.internal', port=80): Max retries exceeded with url: /computeMetadata/v1/instance/id (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f9169432760>: Failed to establish a new connection: [Errno 101] Network is unreachable'))` , should we `continue` on those exception types instead of logging a duplicated message?

Reply:
Trace-backs removed.

> Also since we know the likelihood of seeing HTTPConnectionPool ... should we continue on those exception types instead of logging a duplicated message?

Those exceptions are [caught, logged and not re-raised](https://github.com/canonical/cloud-init/blob/8615c240c2405165d9755b838bc2be5d82ad0c1d/cloudinit/sources/DataSourceGCE.py#L49-L51), thus there is no need to `continue`. We are going to cleanly pass to the next loop cycle if no other errors happen and the connection could have not been established in subsequent retries.